src/operator/deinput.rs

```diff
@@ -0,0 +1,703 @@
+/// object-safe deserializer API ([`erased_serde::Deserializer`]), so that
+/// `DeCollectionHandle` methods don't need to be parameterized by the
+/// deserializer type.
+pub trait DeCollectionHandle: Send {
+    /// Buffer a new insert update.
+    ///
+    /// The `deserializer` argument wraps a single serialized record.
+    /// The value gets deserialized and pushed to the underlying input stream
+    /// handle as an insert update.
+    ///
+    /// Returns an error if deserialization fails, i.e., the serialized
+    /// representation is corrupted or does not match the value type of
+    /// the underlying input stream.
+    ///
+    /// See [`DeZSetHandle`], [`DeSetHandle`], and [`DeMapHandle`]
+    /// documentation for details.
+    fn insert(&mut self, deserializer: &mut dyn ErasedDeserializer) -> Result<(), EError>;
+
+    /// Buffer a new delete update.
+    ///
+    /// The `deserializer` argument wraps a single serialized record whose
+    /// type depends on the underlying input stream: streams created by
+    /// [`Circuit::add_input_zset`](`crate::Circuit::add_input_zset`)
+    /// and [`Circuit::add_input_set`](`crate::Circuit::add_input_set`)
+    /// methods support deletion by value, hence the serialized record must
+    /// match the value type of the stream.  Streams created with
+    /// [`Circuit::add_input_map`](`crate::Circuit::add_input_map`)
+    /// support deletion by key, so the serialized record must match the key
+    /// type of the stream.
+    ///
+    /// The record gets deserialized and pushed to the underlying input stream
+    /// handle as a delete update.
+    ///
+    /// Returns an error if deserialization fails, i.e., the serialized
+    /// representation is corrupted or does not match the value or key
+    /// type of the underlying input stream.
+    ///
+    /// See [`DeZSetHandle`], [`DeSetHandle`], and [`DeMapHandle`]
+    /// documentation for details.
+    fn delete(&mut self, deserializer: &mut dyn ErasedDeserializer) -> Result<(), EError>;
+
+    /// Reserve space for at least `reservation` more updates in the
+    /// internal input buffer.
+    ///
+    /// Reservations are not required but can be used when the number
+    /// of inputs is known ahead of time to reduce reallocations.
+    fn reserve(&mut self, reservation: usize);
+
+    /// Push all buffered updates to the underlying input stream handle.
+    ///
+    /// Flushed updates will be pushed to the stream during the next call
+    /// to [`DBSPHandle::step`](`crate::DBSPHandle::step`).  `flush` can
+    /// be called multiple times between two subsequent `step`s.  Every
+    /// `flush` call adds new updates to the previously flushed updates.
+    ///
+    /// Updates queued after the last `flush` remain buffered in the handle
+    /// until the next `flush` or `clear_buffer` call or until the handle
+    /// is destroyed.
+    fn flush(&mut self);
+
+    /// Clear all buffered updates.
```

Comment:
undo instead of clear?

Reply:
Clear is better, we're not "undoing" anything here

Reply:
Updates haven't been applied yet, only buffered.  This is literally clearing the buffer.